File: rattle.js
'use strict';

(function (global) {
	if (global["WebSocket"]) {

		var call = function (to, data) {
			if (to == "" || to == undefined) return;

			if ("=+@".indexOf(to[0]) != -1) {
				callElement(to, data)
			} else {
				callFunction(to, data)
			}
		}

		var callElement = function (to, data) {
			var element
			switch (to[1]) {
			case "#":
				element = document.getElementById(to.slice(2))
				break;
			default:
				element = document.querySelector(to.slice(1))
				break;
			}

			if (!element) {
				console.error("target element not found: ", to.slice(1))
				return;
			}

			if (to[0] == "@") {
				element.outerHTML = data
			} else {
				if (element.tagName == "INPUT" || element.tagName == "TEXTAREA") {
					element.value = to[0] == "=" ? data : element.value + data
				} else {
					element.innerHTML = to[0] == "=" ? data : element.innerHTML + data
				}
			}
		}

		var callFunction = function (to, data) {
			var func = getFunction(to)
			if (!func) {
				console.error("target function not found: ", to)
				return;
			}

			try {
				data = JSON.parse(data)
				func(data)
			} catch (e) {
				func(data)
			}
		}

		var getFunction = function (functionName) {
			var namespaces = functionName.split(".")
			var func = namespaces.pop()
			var context = window

			for (var i = 0; i < namespaces.length; i++) {
				if (context[namespaces[i]] == undefined) {
					// console.warn("rpc function not found", functionName)
					return undefined;
				}
				context = context[namespaces[i]]
			}
			return context[func];
		}

		var NewConnection = function (addr, debug) {
			this.addr = addr
			this.debug = debug
			this.connected = false
			this.boundary = ""

			this.ws = new WebSocket(addr)
			this.ws.onopen = this.onConnect.bind(this)
			this.ws.onclose = this.onDisconnect.bind(this)
			this.ws.onmessage = this.onMessage.bind(this)

			this.callbacks = {}

			this.connect = connect
			this.disconnect = disconnect

			this.stream = stream
			this.streamNext = streamNext
		}

		NewConnection.prototype = {
			constructor: NewConnection,

			onConnect: function (evt) {
				this.connected = true
				if (this.debug) console.log("rattle: connected")
				if (this.callbacks["onConnect"]) this.callbacks["onConnect"](evt);
			},

			onDisconnect: function (evt) {
				this.connected = false
				if (this.debug) console.log("rattle: disconnected")
				if (this.callbacks["onDisconnect"]) this.callbacks["onDisconnect"](evt);
			},

			onMessage: function (incomingData) {
				var splitted = incomingData.data.split(' '),
					to = splitted[0].trim(),
					data = splitted.slice(1, splitted.length).join(" ").trim()

				if (to == "stream") {
					this.stream()
					return
				}

				if (this.debug) console.log("rattle: Get message to: " + to + " with data length:", data.length)

				if (this.callbacks["onMessage"]) {
					this.callbacks["onMessage"](to, data)
				} else {
					call(to, data)
				}
			},

			event: function (name, callback) {
				this.callbacks[name] = callback;
			},

			send: function (to, data) {
				if (to == undefined || to == "") {
					console.warn("rattle: field 'To'(target function) is not filled")
					return;
				}

				var msg = {}
				msg.to = to
				msg.url = window.location.href
				

				if (data != undefined) {
					msg.data = data
					msg.type = typeof data
				}

				//reconnect if current not connected
				if (this.ws.readyState == 3) {
					this.connect()
				}

				this.ws.send(JSON.stringify(msg) + "\n")

				if (this.debug) console.log("rattle: Send message to: " + to + " with data:", JSON.stringify(data))
			},

			file: function (to, input, userdata) {
				if (input.files.length == 0) {
					console.warn("files not found")
					return
				}

				streamData.to = to
				streamData.url = window.location.href
				streamData.files = input.files
				streamData.userdata = userdata
				streamData.current = 0
				streamData.offset = 0
				streamData.i = 0

				if (this.debug) console.log("rattle: open stream", streamData)

				this.stream()
			},
		} // end newConnection.prototype

		var streamData = {
			to: "",
			url: "",
			files: "",
			userdata: {},
			i: 0,
			offset: 0,
			slicesize: 1024 * 1024
		}

		var streamNext = function (ws) {
			if (streamData.i >= streamData.files.length) {
				return false;
			}
			// console.log(this)
			switch (streamData.offset) {
			case 0:
				// console.log(this)
				var msg = {
					to: streamData.to,
					url: window.location.href,
					type: "stream",
					json: streamData.userdata,
					stream: {
						name: streamData.files[streamData.i].name,
						size: streamData.files[streamData.i].size,
						slicesize: streamData.slicesize
					}
				}

				ws.send(JSON.stringify(msg) + "\n")

				// ws.send(streamData.to + " stream " + JSON.stringify(streamData.userdata) + " " + JSON.stringify({
				// 	name: streamData.files[streamData.i].name,
				// 	size: streamData.files[streamData.i].size,
				// 	slicesize: streamData.slicesize
				// }) + "\n")

				return true;

			case streamData.files[streamData.i].size:
				ws.send("\n{\"type\":\"finish\"}\n")

				streamData.offset = 0
				streamData.i++

				return streamNext(ws);
			default:
				ws.send("\n{\"type\":\"chunk\"}\n")
				return true;
			}
		}

		var stream = function () {
			if (!streamNext(this.ws)) {
				return
			}

			if (this.debug) console.log("rattle: send chunk ", streamData)

			var offsetEnd = streamData.offset + streamData.slicesize
			if (offsetEnd > streamData.files[streamData.i].size) {
				offsetEnd = streamData.files[streamData.i].size
			}

			//reconnect if current not connected
			if (this.ws.readyState == 3) {
				this.connect()
			}

			this.ws.send(streamData.files[streamData.i].slice(streamData.offset, offsetEnd))
			streamData.offset = offsetEnd
		}

		var disconnect = function () {
			if (this.connected) {
				this.ws.close()
			}
		}

		var connect = function () {
			if (this.connected) {
				return
			}
			this.ws = new WebSocket(this.addr)
			this.ws.onopen = this.onConnect.bind(this)
			this.ws.onclose = this.onDisconnect.bind(this)
			this.ws.onmessage = this.onMessage.bind(this)
		}

		global.rattle = {
			NewConnection: NewConnection
		}

	} else {
		console.warn("rattle: WebSockets not supported!")
	}
})(this)

File: rattle.go
package rattle

import (
	"bufio"
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"net/http"

	"github.com/sg3des/bytetree"
	"golang.org/x/net/websocket"
)

var Debug bool

type Rattle struct {
	//Connections map contains all available web socket connections
	Connections *bytetree.Tree

	//Routes contains name and interface link on controller
	Routes *bytetree.Tree

	//events
	onConnect    Handler
	onDisconnect Handler
}

//NewRattle initialize new rattle instance
func NewRattle() *Rattle {
	r := &Rattle{
		Connections: bytetree.NewTree(),
		Routes:      bytetree.NewTree(),
	}

	return r
}

//Handler method type
type Handler func(*Request)

//AddRoute handler method by name
func (r *Rattle) AddRoute(name string, h Handler) {
	r.Routes.GrowLeaf([]byte(name), h)
}

//SetOnConnect bind event onconnect
func (r *Rattle) SetOnConnect(f Handler) {
	r.onConnect = f
}

//SetOnDisconnect bind event onDisconnect
func (r *Rattle) SetOnDisconnect(f Handler) {
	r.onDisconnect = f
}

//Handler return http.Handler with websocket
func (r *Rattle) Handler() http.Handler {
	return websocket.Handler(r.wshandler)
}

//wshandler is handler for websocket connections
func (r *Rattle) wshandler(ws *websocket.Conn) {
	c := r.newСonnection(ws)
	if Debug {
		log.Println("new connection:", string(c.key))
	}

	scanner := bufio.NewScanner(ws)
	for scanner.Scan() && c != nil {
		r.call(c, scanner.Bytes())
	}

	r.Disconnect(c)
}

func (r *Rattle) call(c *Connection, data []byte) {
	req, err := r.parseRequest(data)
	if err != nil {
		if Debug {
			log.Println(err)
		}
		return
	}

	req.conn = c

	if Debug {
		log.Printf("request: %s %s", req.To, req.Data)
	}

	switch req.Type {
	case "stream":
		err = r.stream(req)
	default:
		err = r.request(req)
	}

	if err != nil && Debug {
		log.Println(err)
	}
}

//Connection is main struct contains websocket.Conn
type Connection struct {
	key []byte
	w   *bytes.Buffer
	ws  *websocket.Conn
}

func (r *Rattle) newСonnection(ws *websocket.Conn) *Connection {
	key := []byte(fmt.Sprintf("%p", ws))
	if v, ok := r.Connections.LookupLeaf(key); ok {

		return v.(*Connection)
	}

	c := &Connection{
		key: key,
		ws:  ws,
		w:   bytes.NewBuffer([]byte{}),
	}

	r.Connections.GrowLeaf(key, c)

	if r.onConnect != nil {
		go r.onConnect(&Request{conn: c})
	}

	return c
}

//Disconnect end the current connection
func (r *Rattle) Disconnect(c *Connection) {
	if c != nil {
		if r.onDisconnect != nil {
			r.onDisconnect(&Request{conn: c})
		}

		r.Connections.CutLeaf(c.key)
		c.ws.Close()
		c = nil
	}
}

type Request struct {
	conn *Connection

	To     string
	Type   string
	URL    string `json:"url"`
	Data   json.RawMessage
	Stream json.RawMessage

	File *File
}

//File structure
type File struct {
	Name string `json:"name"`
	Size int    `json:"size"`

	Buffer *bytes.Buffer

	SliceSize int `json:"slicesize"`
}

func (req *Request) DecodeTo(v interface{}) error {
	return json.Unmarshal(req.Data, v)
}

//parsemsg parse []byte message to type Message
func (r *Rattle) parseRequest(buf []byte) (req *Request, err error) {
	err = json.Unmarshal(buf, &req)
	return
}

//Request is main function, takes connection and raw incoming message
func (r *Rattle) request(req *Request) error {
	v, ok := r.Routes.LookupLeaf([]byte(req.To))
	if !ok {
		return errors.New("404 page not found")
	}

	v.(Handler)(req)
	return nil
}

func (r *Rattle) stream(req *Request) error {
	req.conn.ws.Write([]byte("stream --"))

	err := json.Unmarshal(req.Stream, &req.File)
	if err != nil {
		if Debug {
			log.Println("rattle: failed unmarshal file header:", err)
		}
		return err
	}

	req.File.Buffer = bytes.NewBuffer([]byte{})
	log.Println(req.File)

STREAM:
	for {
		line := make([]byte, req.File.SliceSize)
		n, err := req.conn.ws.Read(line)
		if err != nil {
			if Debug {
				log.Println(err)
			}
			return err

		}
		line = line[:n]

		rr, err := r.parseRequest(line)
		// if err != nil {
		// 	if Debug {
		// 		log.Println(err)
		// 	}
		// 	continue
		// }

		if err == nil {
			switch rr.Type {
			case "chunk":
				req.conn.ws.Write([]byte("stream --"))
				continue STREAM
			case "finish":
				break STREAM
			}
		}

		if !bytes.Equal(line, []byte("\n")) {
			req.File.Buffer.Write(line)
		}
		// }

		if err != nil {
			break STREAM
		}
	}

	return r.request(req)
}

//Message type:
//  To - contains name of called function, must be filled!
//  Data may contains payload in json format - for backend, or json,html or another for frontend, not necessary.
type Message struct {
	conn *Connection
	To   []byte
	Data []byte
}

//NewMessage create answer message
func (req *Request) NewMessage(to string, data []byte) *Message {
	msg := &Message{
		To:   []byte(to),
		conn: req.conn,
	}

	if len(data) > 0 {
		msg.Data = data
	}

	return msg
}

//Bytes convert Message type to []byte, for write to socket
func (m *Message) Bytes() (bmsg []byte) {
	buf := bytes.NewBuffer(bmsg)

	buf.Write(m.To)
	buf.WriteRune(' ')
	buf.Write(m.Data)

	bmsg = buf.Bytes()
	// bmsg = regexp.MustCompile("(?m)\n+$").ReplaceAll(bmsg, []byte("\n"))

	return
}

//Send message to connection
func (m *Message) Send() error {
	if m.conn == nil || m.conn.ws == nil {
		err := errors.New("connection is nil")
		if Debug {
			log.Println(err)
		}
		return err
	}
	_, err := m.conn.ws.Write(m.Bytes())

	if Debug {
		log.Printf("send message to: %s with data: %s", m.To, m.Data)
	}

	return err
}

//Broadcast send one message for all available connections(users)
func (r *Rattle) Broadcast(m *Message) {
	for _, v := range r.Connections.PickAllLeafs() {
		v.(Connection).ws.Write(m.Bytes())
	}
}
